lib/IDE/CodeCompletionResultBuilder.h

```diff
@@ -86,15 +51,21 @@ class CodeCompletionResultBuilder {
   llvm::PointerUnion<const ModuleDecl *, const clang::Module *>
       CurrentModule;
   ExpectedTypeContext declTypeContext;
-  CodeCompletionResultTypeRelation ExpectedTypeRelation =
-      CodeCompletionResultTypeRelation::Unknown;
   bool Cancelled = false;
   ContextFreeNotRecommendedReason ContextFreeNotRecReason =
       ContextFreeNotRecommendedReason::None;
   ContextualNotRecommendedReason ContextualNotRecReason =
       ContextualNotRecommendedReason::None;
   StringRef BriefDocComment;
 
+  /// The result type that this completion item produces.
+  CodeCompletionResultType ResultType = CodeCompletionResultType::unknown();
+
+  /// The context in which this completion item is used. Used to compute the
+  /// type relation to \c ResultType.
+  ExpectedTypeContext TypeContext;
```

Comment:
Copying `ExpectedTypeContext` is not trivial. Could you use a reference?

Reply:
Not super easily. If `setResultTypes` was never called, we don’t have a `TypeContext`. Currently, it just defaults to an empty `ExpectedTypeContext`. We could make this a pointer but this also means that we need to accept a pointer in the `CodeCompletionResult` initializer, which, I think, makes that API worse.

Reply:
> We could make this a pointer but this also means that we need to accept a pointer in the CodeCompletionResult initializer

Can we dereference the pointer when calling `CodeCompletionResult` constructor? 

---
That being said, I now feel like the ideal solution is that `CodeCompletionResultBuilder` only build `ContextFreeCodeCompletionResult`. Then, after collecting all context free results including cached results, we can loop through them to construct contextual `CodeCompletionResult`s using `ExpectedTypeContext` and `DeclContext`.

Reply:
Ah, I just thought about it again and figured out it’s better to pass `ExpectedTypeContext` around using pointers instead of references anyway to match the fact that we are also using `DeclContext *` and not `DeclContext &`.